### ISSUE TITLE
Allow user to indicate where the TLD file is located

### ISSUE THREAD

**Issue by Yomguithereal** [barseghyanartur/tld#62]
In some of my use cases, being a web crawler for research, I need to have a fixed definition of the TLD across time (sometimes several years). Until then, we just passed a fixed version of the tlds to our internal functions. But now we are migrating to your lib and would like to be able to do the same.

I therefore propose:

1. To get rid of the `conf` module
2. To make the `update_tld_names` function simply take the target url and path location as argument and give it your defaults
3. To add a function such as `set_tld_names_location` or something of the kind, taking a custom path on the user's computer.

What do you think?

**Comment by Yomguithereal** [comment]
Related to #12

**Comment by barseghyanartur** [comment]
@Yomguithereal:

I think we could easily add a ``tld_names_local_path`` argument to some functions in order to solve your issue.

As a result, you would be able to have as many local tld data files as you wish.

```python
tld_obj = get_tld('http://google.com', as_object=True, tld_names_local_path='/usr/tlds/effective_tld_names-2019-10-27.dat.txt')
```

And in some other case:

```python
tld_obj = get_tld('http://google.com', as_object=True, tld_names_local_path='/usr/tlds/effective_tld_names-2018-05-24.dat.txt')
```

That would be very easy to implement. If you're think this would solve your issue, you're welcome to submit a PR.

Alternatively, we could introduce a concept of TLD providers. Mozilla would be one of them. Base provider would be implement most of the abstract staff and would contain links to both: the external TLD source (from which the updates would have to be downloaded) and the local TLD source. So, for instance, if you would want to maintain your own custom TLD version, you would register your own tld provider and call it like:

```python
from tld.base import BaseProvider
from tld.registry import register_provider

class MozillaCustomProvider(BaseProvider):
    uid = 'custom_mozilla'
    # The rest of the methods and properties to follow

tld_obj = get_tld('http://google.com', as_object=True, provider='custom_mozilla')
```

This would certainly be more work to do. 

Please, let me know what you think.

**Comment by barseghyanartur** [comment]
@Yomguithereal:

Ping. :) Did you have time to think on this?

**Comment by Yomguithereal** [comment]
Having this option on the `get_tld` method could incur some kind of slowdown? Does this mean we need to cache potentially several tries at once? I like this because it means a same script can rely on more than a single tld dat file but I am worried about perf. Another way to go could also to have a class exposing the methods like `get_tld` and taking paths & urls as instantiation options and export a default one as the current functions?

**Comment by Yomguithereal** [comment]
```python
class MozillaCustomProvider(BaseProvider):
    uid = 'custom_mozilla'
    # The rest of the methods and properties to follow

tld_obj = get_tld(
    'http://google.com', 
    as_object=True, 
    provider='custom_mozilla'
)
```

I dislike the provider as string. I would rather have something like:

```python
class MozillaCustomProvider(BaseProvider):
    uid = 'custom_mozilla'
    # The rest of the methods and properties to follow

provider = CustomProvider

tld_obj = get_tld(
    'http://google.com', 
    as_object=True, 
    provider=provider
)
```

**Comment by barseghyanartur** [comment]
@Yomguithereal:

Providers would be more difficult to implement. I think provider might be a string (which then would be used to pull a correspondent class from registry) or a python class (inherited from BaseProvider).

As of custom tld dat file path `tld_names_local_path`, `tld_names` should become a dict where keys would be the paths to `dat` files and values would be the Trie objects. There would be a tiny memory increase, but I don't think it would affect performance anyhow (unless user would use way to many different `dat` files to eat up the whole memory).

**Comment by barseghyanartur** [comment]
@Yomguithereal:

Could you evaluate [this](https://github.com/barseghyanartur/tld/tree/feature/62-Allow-user-to-indicate-where-the-TLD-file-is-located)?

See the [doc additions as well](https://github.com/barseghyanartur/tld/tree/feature/62-Allow-user-to-indicate-where-the-TLD-file-is-located#custom-list-of-tld-names).

**Comment by Yomguithereal** [comment]
Hello @barseghyanartur. Looks good to me. However I have some concerns:

1. What if you need to release some trie from memory (this is an unlikely use case where someone would use many different lists for some reasons and would like to release some lists from memory some times to avoid mem saturation).
2. In my use case (the library [ural](https://github.com/medialab/ural)), I will need to enable users to tell me where the list is. So I have two solutions (which are both possible with your implementation):
    1. Add a the same kwarg `tld_names_local_path` to my relevant functions.
    2. Provide a way to swap the path at library level (something like: `ural.set_tld_names_local_path`), which I can implement using partials etc. so your solution still works for me.

@boogheta what's your take on this?

**Comment by barseghyanartur** [comment]
@Yomguithereal:

Regarding (1):

We can make a function for it or alter ``tld.utils.reset_tld_names`` to take an optional ``tld_names_local_path`` argument to free up that one only.

As of (2):

Sounds reasonable.

If everyone is happe, I'll make changes regarding freeing up resources and then make a release.

**Comment by boogheta** [comment]
Hello,
To answer @Yomguithereal's question, in Hyphe's usecase, we would be more interested in having the possibility to pass not a file path or url where to collect the TLDs, but rather an object argument containing for instance the tree of TLDs as it was at the time it was collected when we first instantiated a new corpus and stored it locally in a DB as configuration of this corpus.
cf https://github.com/medialab/hyphe/blob/master/hyphe_backend/lib/tlds.py

**Comment by barseghyanartur** [comment]
@boogheta:

Fine. Ok, so some sort of a frozen object. Do you think you could turn it into a ``Trie`` object (as defined in ``tld.utils``? I'm thinking of pickling the result and adding a function that would simply add the compiled ``Trie`` into the ``tld.utils.tld_names``.

**Comment by Yomguithereal** [comment]
@barseghyanartur I can also upgrade the Trie to output a lean binary or string representation if we want to avoid the size cost of pickling.

**Comment by barseghyanartur** [comment]
@Yomguithereal:

Yep. I want to finish what's done and make a release. 
Could you then prepare a PR?

**Comment by Yomguithereal** [comment]
Yes but I wont be able to do it soon. Don’t wait for me to ship :)

**Comment by barseghyanartur** [comment]
0.10 is released. I'm leaving this issue open for the second part.

**Comment by boogheta** [comment]
So far, to store my dictionary tree into my DB, I was using msgpack, which is like pickle but way faster and more efficient.
I'd go along with doing the same with a different structure built on this lib side.

**Comment by barseghyanartur** [comment]
@boogheta:

Good to hear that. However, don't bother yet, 'cause I'm introducing parser classes to make things more flexible than they were. Your additions will be very much welcome once it's done.

**Comment by barseghyanartur** [comment]
@Yomguithereal:

Could you evaluate [this](https://github.com/barseghyanartur/tld/tree/feature/62-Introduce-parser-classes)?

See the [doc additions as well](https://github.com/barseghyanartur/tld/tree/feature/62-Introduce-parser-classes#custom-tld-parsers).

I noticed you're using data from publicsuffix.org. You could eventually make a parser for publicsuffix.org and come up with PR, if you like.

@boogheta:

Could you check the changes as well? If you think compiled dictionary tree (using pickle, msgpack or whatever) adds significant load performance (even if first time only), you're welcome to submit a PR (or eventually, consult me on details so that I could add it myself). The way I see it is that it should be possible to compile complete TLD trie of each parser and dump it into the `res` directory. I guess it could be something similar to ``tld.utils.update_tld_names_cli``, that will produce a file (that later on can be added to the repository).

**Comment by boogheta** [comment]
Hi and thanks for the followup!
To be more clear: the need within our Hyphe project is to collect TLDs at once when we start a user's project and then store it in database (among other settings of each project, within mongo, but this does not change anything) so that it can be reused later on for this project, but also so that we can send the corresponding list of TLDs to a javascript client which uses it its own way.
So what I would really like the best would be to be able to feed get_tld with a string serialization of the list of TLDs.
One option would be to get it under the form of a list, similar to the content of publicsuffix.org, although I understand that caching the trie for performance becomes more complex when starting from a huge string and not just a file path...
So maybe a serialization of the trie would be the only solution, but then this means I need to be able to regenerate the list or the tree of TLDs from this serialization of the trie (which sounds doable).
cc @Yomguithereal

**Comment by barseghyanartur** [comment]
Implemented in 0.11. I'm opening another issue for additions mentioned by you.
